src/utils.c

```diff
@@ -37,43 +37,28 @@ void rv_gettimeofday(struct timeval *tv)
         (void) mach_timebase_info(&info);
     /* Hope that the multiplication doesn't overflow. */
     uint64_t nsecs = mach_absolute_time() * info.numer / info.denom;
-    int32_t tv_sec = nsecs / 1e9;
-    int32_t tv_usec = (nsecs / 1e3) - (tv_sec * 1e6);
+    *tv_sec = nsecs / 1e9;
+    *tv_usec = (nsecs / 1e3) - (tv_sec * 1e6);
 #else /* low resolution timer */
     clock_t t = clock();
-    int32_t tv_sec = t / CLOCKS_PER_SEC;
-    int32_t tv_usec = (t % CLOCKS_PER_SEC) * (1000000 / CLOCKS_PER_SEC);
+    *tv_sec = t / CLOCKS_PER_SEC;
+    *tv_usec = (t % CLOCKS_PER_SEC) * (1e6 / CLOCKS_PER_SEC);
 #endif
+}
 
+void rv_gettimeofday(struct timeval *tv)
+{
+    int32_t tv_sec, tv_usec;
+    get_time_info(&tv_sec, &tv_usec);
     tv->tv_sec = tv_sec;
     tv->tv_usec = tv_usec;
 }
 
 /* TODO: clarify newlib internals */
 void rv_clock_gettime(struct timespec *tp)
 {
-#if defined(HAVE_POSIX_TIMER)
-    struct timespec t;
-    clock_gettime(CLOCKID, &t);
-    int32_t tv_sec = t.tv_sec;
-    int32_t tv_msec = t.tv_nsec / 1e6; /* resolution (ms) */
-#elif defined(HAVE_MACH_TIMER)
-    static mach_timebase_info_data_t info;
-    /* If it is the first time running, obtain the timebase. Using denom == 0
-     * indicates that sTimebaseInfo is uninitialized.
-     */
-    if (info.denom == 0)
-        (void) mach_timebase_info(&info);
-    /* Hope that the multiplication doesn't overflow. */
-    uint64_t nsecs = mach_absolute_time() * info.numer / info.denom;
-    int32_t tv_sec = nsecs / 1e9;
-    int32_t tv_msec = (nsecs / 1e6) - (tv_sec * 1e9);
-#else /* low resolution timer */
-    clock_t t = clock();
-    int32_t tv_sec = t / CLOCKS_PER_SEC;
-    int32_t tv_msec = ((t / 1000) % CLOCKS_PER_SEC) * (1e6 / CLOCKS_PER_SEC);
-#endif
-
+    int32_t tv_sec, tv_usec;
+    get_time_info(&tv_sec, &tv_usec);
     tp->tv_sec = tv_sec;
-    tp->tv_nsec = tv_msec;
+    tp->tv_nsec = tv_usec / 1000; /* Transfer to microseconds */
```

Comment:
The statement appears somewhat unusual. When I initially worked on incorporating newlib support, I encountered some confusion regarding the time scale and units. Could you provide clarification regarding the implementation and expectations related to newlib support?

Reply:
The reason for this statement is that when I was reviewing commit [686554c](https://github.com/visitorckw/rv32emu/commit/686554c58e00de03f18957b1310d59d2b2346b74), I noticed that Newlib was manipulating time with millisecond resolution, even though clock_gettime expects nanoseconds in the timespec struct. However, I did not investigate this further at the time. Perhaps I can take a closer look to understand the underlying reasons.

Reply:
Add TODO/FIXME comments and submit another pull request(s) when you are ready to resolve the newlib issue.

Reply:
I believe the comment on line 57 of the code, /* TODO: clarify newlib internals */, is referring to this issue.

Reply:
> I believe the comment on line 57 of the code,`/* TODO: clarify newlib internals */`, is referring to this issue.

Exactly, you can describe more this time.